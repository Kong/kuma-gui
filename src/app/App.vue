<template>
  <RouteView>
    <AppLoadingBar v-if="store.state.globalLoading || route.name === undefined" />

    <template v-else>
      <AppHeader v-if="!isWizard" />

      <div v-if="route.meta.onboardingProcess">
        <router-view />
      </div>

      <div
        v-else
        class="app-content-container"
      >
        <AppSidebar v-if="!isWizard" />

        <AppView>
          <AppErrorMessage
            v-if="shouldShowAppError"
            data-testid="app-error"
          />

          <NotificationManager v-if="!isWizard && shouldShowNotificationManager" />

          <AppOnboardingNotification v-if="!isWizard && shouldShowOnboardingNotification" />

          <router-view
            :key="routeKey"
            v-slot="{ Component }"
          >
            <transition
              mode="out-in"
              name="fade"
            >
              <div
                :key="(route.name as string)"
                class="transition-root"
              >
                <component
                  :is="Component"
                  :data="props.data"
                />
              </div>
            </transition>
          </router-view>
        </AppView>
      </div>
    </template>
  </RouteView>
</template>

<script lang="ts" setup>
import { computed, watch } from 'vue'
import { useRoute } from 'vue-router'

import AppView from '@/app/application/components/app-view/AppView.vue'
import RouteView from '@/app/application/components/route-view/RouteView.vue'
import {
  useAppSidebar,
  useAppHeader,
  useAppLoadingBar,
  useAppErrorMessage,
  useNotificationManager,
  useAppOnboardingNotification,
} from '@/components'
import { useStore } from '@/store/store'
const props = defineProps({
  data: {
    type: Object,
    required: false,
    default: undefined,
  },
})
const [
  AppSidebar,
  AppHeader,
  AppLoadingBar,
  AppErrorMessage,
  NotificationManager,
  AppOnboardingNotification,
] = [
  useAppSidebar(),
  useAppHeader(),
  useAppLoadingBar(),
  useAppErrorMessage(),
  useNotificationManager(),
  useAppOnboardingNotification(),
]
const store = useStore()
const route = useRoute()

/**
 * The `router-view`’s `key` attribute value.
 *
 * Set to the current `route.path` to trigger an explicit re-render via Vue’s `key` mechanism when navigating between routes that only change in dynamic path segments while matching the same route definition.
 *
 * **Example**:
 *
 * From: /mesh/default/services/backend
 * To: /mesh/default/services/ingress
 *
 * Both routes resolve to the same route definition and the router’s default behavior is to not re-render the component in such navigations.
 */
const routeKey = computed(() => route.path)
const isWizard = computed(() => route.meta.isWizard === true)
const shouldShowAppError = computed(() => store.getters.shouldShowAppError)
const shouldShowNotificationManager = computed(() => store.getters.shouldShowNotificationManager)
const shouldShowOnboardingNotification = computed(() => store.getters.shouldShowOnboardingNotification)

watch(() => isWizard.value, setIsWizardPageClass, { immediate: true })

/**
 * Adds a class for wizard pages to the body element. This is used to control certain layout aspects of the app.
 */
function setIsWizardPageClass(isWizard: boolean) {
  const hasClass = document.body.classList.contains('is-wizard-page')

  if (isWizard && !hasClass) {
    document.body.classList.add('is-wizard-page')
  } else if (!isWizard && hasClass) {
    document.body.classList.remove('is-wizard-page')
  }
}
</script>

<style lang="scss" scoped>
body:not(.is-wizard-page) .app-content-container {
  padding-top: var(--AppHeaderHeight, initial);
  display: grid;
  grid-template-columns: var(--AppSidebarWidth) 1fr;
}

body:not(.is-wizard-page) .app-main-content {
  padding: var(--AppGap);
}
</style>
