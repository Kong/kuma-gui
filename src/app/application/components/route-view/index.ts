export const ROUTE_VIEW_ROOT = Symbol('route-view-root')
export const ROUTE_VIEW_PARENT = Symbol('route-view-parent')
