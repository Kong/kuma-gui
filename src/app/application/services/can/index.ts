export type Can = (str: string, obj?: any) => boolean
export default (features: Record<string, (...args: any[]) => boolean>) => {
  const can: Can = (str, obj) => {
    const feature = features[str]
    if (typeof feature !== 'undefined') {
      return features[str](can, obj)
    }
    return false
  }
  return can
}
