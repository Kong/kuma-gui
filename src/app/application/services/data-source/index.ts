import CallableEventSource from './CallableEventSource'
import type { Creator, Destroyer } from './DataSourcePool'
export type { DataSourceResponse } from './DataSourcePool'

type Configuration = {
  interval?: number
  retry?: (e: unknown) => Promise<void> | undefined
}
type RetryingEventSource = CallableEventSource<Configuration>
type Hideable = EventTarget & { hidden: boolean }

export const getSource = (doc: Hideable) => {
  return (cb: (source: RetryingEventSource) => Promise<unknown>, config: Configuration = {}) => {
    let attempts = 0
    let iterations = 0
    return new CallableEventSource<Configuration>(async function * (this: RetryingEventSource) {
      const self = this
      while (true) {
        self.readyState = 1
        // this this isn't the first call then we should wait before calling again
        if (iterations > 0) {
          await new Promise((resolve) => setTimeout(resolve, self.configuration.interval ?? 1000))
        }
        if (attempts > 0 || iterations > 0) {
          // if the document/browser tab is hidden then wait for it to regain
          // focus but, for the first call (if we aren't erroring) we probably
          // still want to send of the request, so then at least when you come
          // back you immediately see data
          if (doc.hidden) {
            await new Promise((resolve) => {
              doc.addEventListener('visibilitychange', resolve, { once: true })
            })
          }
        }
        let res
        try {
          res = cb(self)
          // if we aren't polling then immediately close after calling
          if (typeof self.configuration.interval === 'undefined') {
            self.close()
          }
          // only increase iterations if we didn't error
          iterations++
          // return the result
          yield res
        } catch (e) {
          // if retry is configured await it before entering the loop again to
          // try again
          // TODO(jc): we should probably pass through attempts and maybe other
          // things here
          const retry = self.configuration?.retry?.(e)
          if (typeof retry?.then === 'function') {
            // make sure we never mistakenly retry sooner than 1s
            await Promise.all([retry, new Promise(resolve => setTimeout(resolve, 1000))])
            attempts++
          } else {
            throw e
          }
        }
      }
    }, config)
  }
}
export type Source = ReturnType<typeof getSource>

// its fine to not wait for an unfocussed tab for promise returning sources
const source = getSource(new (class extends EventTarget {hidden = false})())

export const create: Creator = (src, router) => {
  const [path, query] = src.split('?')
  const queryParams = new URLSearchParams(query)
  // use the router to find which function to call
  const route = router.match(path)
  const params = {
    ...{
      offset: parseInt(queryParams.get('offset') || '0'),
      size: parseInt(queryParams.get('size') || '0'),
      page: parseInt(queryParams.get('page') || '0'),
      search: queryParams.get('search') || '',
    },
    ...route.params,
  }
  try {
    // TODO(jc) Once we remove all the source.closes in the sources.ts files the
    // second argument here can go
    const init = route.route(params, { close: () => {} })
    if (init instanceof CallableEventSource) {
      return init
    } else {
      return source(() => Promise.resolve(init))
    }
  } catch (e) {
    return source(() => Promise.reject(e))
  }
}
export const destroy: Destroyer = (_src, source) => {
  if (source) {
    source.close()
  }
}

export default {
  create,
  destroy,
}
