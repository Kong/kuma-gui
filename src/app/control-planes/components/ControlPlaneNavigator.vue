<template>
  <AppNavigator
    data-testid="control-planes-navigator"
  >
    <RouterLink
      :class="{ 'is-active': [$route.name].concat($route.matched.map(item => item.name)).some((item) => item === 'home') }"
      :to="{
        name: 'home',
      }"
    >
      Home
    </RouterLink>
  </AppNavigator>
</template>
<script lang="ts" setup>
import AppNavigator from '@/app/application/components/app-navigator/AppNavigator.vue'
</script>
