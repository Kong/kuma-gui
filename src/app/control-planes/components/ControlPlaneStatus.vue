<template>
  <KCard>
    <div class="card-header">
      <div class="card-title">
        <h2>{{ t('main-overview.detail.health.title') }}</h2>
      </div>
    </div>

    <div class="columns">
      <ResourceStatus
        v-if="props.canUseZones"
        :total="props.globalInsight.zones.controlPlanes.total"
        :online="props.globalInsight.zones.controlPlanes.online"
        data-testid="zone-control-planes-status"
      >
        <template #icon>
          <img src="@/assets/images/icon-location-on.svg?url">
        </template>

        <template #title>
          {{ t('main-overview.detail.health.zone_control_planes') }}
        </template>
      </ResourceStatus>

      <ResourceStatus
        :total="props.globalInsight.meshes.total"
        data-testid="meshes-status"
      >
        <template #icon>
          <img src="@/assets/images/icon-circles-ext.svg?url">
        </template>

        <template #title>
          {{ t('main-overview.detail.health.meshes') }}
        </template>
      </ResourceStatus>

      <ResourceStatus
        :total="props.globalInsight.services.internal.total"
        :online="props.globalInsight.services.internal.online"
        data-testid="services-status"
      >
        <template #icon>
          <img src="@/assets/images/icon-wifi-tethering.svg?url">
        </template>

        <template #title>
          {{ t('main-overview.detail.health.services') }}
        </template>
      </ResourceStatus>

      <ResourceStatus
        :total="props.globalInsight.dataplanes.standard.total"
        :online="props.globalInsight.dataplanes.standard.online"
        data-testid="data-plane-proxies-status"
      >
        <template #icon>
          <img src="@/assets/images/icon-wifi-tethering.svg?url">
        </template>

        <template #title>
          {{ t('main-overview.detail.health.data_plane_proxies') }}
        </template>
      </ResourceStatus>
    </div>
  </KCard>
</template>

<script lang="ts" setup>
import { useI18n } from '@/app/application'
import ResourceStatus from '@/app/common/ResourceStatus.vue'
import type { GlobalInsight } from '@/types/index.d'

const { t } = useI18n()

const props = defineProps<{
  globalInsight: GlobalInsight
  canUseZones: boolean
}>()
</script>

<style lang="scss" scoped>
.card-header {
  margin-bottom: $kui-space-50;
  min-height: 42px;
}
</style>
