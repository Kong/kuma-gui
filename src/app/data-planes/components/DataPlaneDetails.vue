<template>
  <TabsWidget
    :tabs="filteredTabs"
    initial-tab-override="overview"
  >
    <template #tabHeader>
      <div>
        <h3>
          DPP: {{ dataPlane.name }}
        </h3>
      </div>

      <div>
        <EntityURLControl
          :name="dataPlane.name"
          :mesh="dataPlane.mesh"
        />
      </div>
    </template>

    <template #overview>
      <LabelList>
        <div>
          <ul>
            <li
              v-for="(value, prop) in processedDataPlane"
              :key="prop"
            >
              <h4>{{ prop }}</h4>

              <div v-if="prop === 'status' && typeof value !== 'string'">
                <div
                  class="entity-status"
                  :class="{
                    'is-offline': value.status.toLowerCase() === 'offline',
                    'is-online': value.status.toLowerCase() === 'online',
                    'is-degraded': value.status.toLowerCase() === 'partially degraded',
                    'is-not-available': value.status.toLowerCase() === 'not available',
                  }"
                >
                  <span>{{ value.status }}</span>
                </div>

                <div
                  v-for="(reason, index) in value.reason"
                  :key="index"
                  class="reason"
                >
                  {{ reason }}
                </div>
              </div>

              <div v-else>
                {{ value }}
              </div>
            </li>
          </ul>
        </div>

        <div>
          <template v-if="dataPlaneTags.length > 0">
            <h4>Tags</h4>

            <p>
              <span
                v-for="(tag, index) in dataPlaneTags"
                :key="index"
                class="tag-cols"
              >
                <span>
                  {{ tag.label }}:
                </span>

                <span>
                  {{ tag.value }}
                </span>
              </span>
            </p>
          </template>

          <template v-if="dataPlaneVersions">
            <h4>Versions</h4>

            <p>
              <span
                v-for="(version, dependency) in dataPlaneVersions"
                :key="dependency"
                class="tag-cols"
              >
                <span>
                  {{ dependency }}:
                </span>

                <span>
                  {{ version }}
                </span>
              </span>
            </p>
          </template>
        </div>
      </LabelList>

      <div class="config-wrapper">
        <YamlView
          id="code-block-data-plane"
          :content="rawDataPlane"
          is-searchable
        />
      </div>
    </template>

    <template #insights>
      <StatusInfo :is-empty="insightSubscriptions.length === 0">
        <KCard border-variant="noBorder">
          <template #body>
            <AccordionList :initially-open="0">
              <AccordionItem
                v-for="(insight, key) in insightSubscriptions"
                :key="key"
              >
                <template #accordion-header>
                  <SubscriptionHeader :details="insight" />
                </template>

                <template #accordion-content>
                  <SubscriptionDetails
                    :details="insight"
                    is-discovery-subscription
                  />
                </template>
              </AccordionItem>
            </AccordionList>
          </template>
        </KCard>
      </StatusInfo>
    </template>

    <template #dpp-policies>
      <DataplanePolicies
        :mesh="dataPlane.mesh"
        :dpp-name="dataPlane.name"
      />
    </template>

    <template #xds-configuration>
      <EnvoyData
        data-path="xds"
        :mesh="dataPlane.mesh"
        :dpp-name="dataPlane.name"
        query-key="envoy-data-data-plane"
      />
    </template>

    <template #envoy-stats>
      <EnvoyData
        data-path="stats"
        :mesh="dataPlane.mesh"
        :dpp-name="dataPlane.name"
        query-key="envoy-data-data-plane"
      />
    </template>

    <template #envoy-clusters>
      <EnvoyData
        data-path="clusters"
        :mesh="dataPlane.mesh"
        :dpp-name="dataPlane.name"
        query-key="envoy-data-data-plane"
      />
    </template>

    <template #mtls>
      <LabelList>
        <ul v-if="mtlsData !== null">
          <li
            v-for="(val, key) in mtlsData"
            :key="key"
          >
            <h4>{{ val.label }}</h4>

            <p>
              {{ val.value }}
            </p>
          </li>
        </ul>

        <KAlert
          v-else
          appearance="danger"
        >
          <template #alertMessage>
            This data plane proxy does not yet have mTLS configured â€”
            <a
              :href="`https://kuma.io/docs/${kumaDocsVersion}/documentation/security/#certificates`"
              class="external-link"
              target="_blank"
            >
              Learn About Certificates in {{ PRODUCT_NAME }}
            </a>
          </template>
        </KAlert>
      </LabelList>
    </template>

    <template #warnings>
      <WarningsWidget :warnings="warnings" />
    </template>
  </TabsWidget>
</template>

<script lang="ts" setup>
import { computed, ref, PropType } from 'vue'
import { KCard, KAlert } from '@kong/kongponents'

import {
  Compatibility,
  DataPlane,
  DataPlaneOverview,
} from '@/types'
import {
  compatibilityKind,
  COMPATIBLE,
  dpTags,
  getStatus,
  getVersions,
  INCOMPATIBLE_WRONG_FORMAT,
  INCOMPATIBLE_ZONE_CP_AND_KUMA_DP_VERSIONS,
  parseMTLSData,
} from '@/dataplane'
import { KUMA_ZONE_TAG_NAME, PRODUCT_NAME } from '@/consts'
import { stripTimes } from '@/helpers'
import { useStore } from '@/store/store'
import AccordionItem from '@/components/Accordion/AccordionItem.vue'
import AccordionList from '@/components/Accordion/AccordionList.vue'
import DataplanePolicies from '@/components/DataplanePolicies/DataplanePolicies.vue'
import EntityURLControl from '@/components/Utils/EntityURLControl.vue'
import EnvoyData from '@/components/EnvoyData/EnvoyData.vue'
import LabelList from '@/components/Utils/LabelList.vue'
import StatusInfo from '@/components/Utils/StatusInfo.vue'
import SubscriptionDetails from '@/views/Entities/components/SubscriptionDetails.vue'
import SubscriptionHeader from '@/views/Entities/components/SubscriptionHeader.vue'
import TabsWidget from '@/components/Utils/TabsWidget.vue'
import WarningsWidget from '@/views/Entities/components/WarningsWidget.vue'
import YamlView from '@/app/common/YamlView.vue'

const store = useStore()

const props = defineProps({
  dataPlane: {
    type: Object as PropType<DataPlane>,
    required: true,
  },

  dataPlaneOverview: {
    type: Object as PropType<DataPlaneOverview>,
    required: true,
  },
})

const tabs = [
  {
    hash: '#overview',
    title: 'Overview',
  },
  {
    hash: '#insights',
    title: 'DPP Insights',
  },
  {
    hash: '#dpp-policies',
    title: 'Policies',
  },
  {
    hash: '#xds-configuration',
    title: 'XDS Configuration',
  },
  {
    hash: '#envoy-stats',
    title: 'Stats',
  },
  {
    hash: '#envoy-clusters',
    title: 'Clusters',
  },
  {
    hash: '#mtls',
    title: 'Certificate Insights',
  },
  {
    hash: '#warnings',
    title: 'Warnings',
  },
]

const warnings = ref<Compatibility[]>([])

const processedDataPlane = computed(() => {
  const { type, name, mesh } = props.dataPlane
  const status = getStatus(props.dataPlane, props.dataPlaneOverview.dataplaneInsight)

  return {
    type,
    name,
    mesh,
    status,
  }
})

const dataPlaneTags = computed(() => dpTags(props.dataPlane))
const dataPlaneVersions = computed(() => getVersions(props.dataPlaneOverview.dataplaneInsight))
const rawDataPlane = computed(() => stripTimes(props.dataPlane))
const mtlsData = computed(() => parseMTLSData(props.dataPlaneOverview))
const insightSubscriptions = computed(() => {
  const subscriptions = Array.from(props.dataPlaneOverview.dataplaneInsight.subscriptions)

  subscriptions.reverse()

  return subscriptions
})

const kumaDocsVersion = computed(() => {
  const storedVersion = store.getters.getKumaDocsVersion

  return storedVersion !== null ? storedVersion : 'latest'
})

const filteredTabs = computed(() => warnings.value.length === 0 ? tabs.filter((tab) => tab.hash !== '#warnings') : tabs)

function setWarnings() {
  const subscriptions = props.dataPlaneOverview.dataplaneInsight.subscriptions

  if (subscriptions.length === 0 || !('version' in subscriptions[0])) {
    return
  }

  const version = subscriptions[0].version

  if (version && version.kumaDp && version.envoy) {
    const compatibility = compatibilityKind(version)

    if (compatibility.kind !== COMPATIBLE && compatibility.kind !== INCOMPATIBLE_WRONG_FORMAT) {
      warnings.value.push(compatibility)
    }
  }

  const isMulticluster = store.getters['config/getMulticlusterStatus']

  if (isMulticluster && version) {
    const tags = dpTags(props.dataPlane)
    const zoneTag = tags.find(tag => tag.label === KUMA_ZONE_TAG_NAME)

    if (zoneTag && typeof version.kumaDp.kumaCpCompatible === 'boolean' && !version.kumaDp.kumaCpCompatible) {
      warnings.value.push({
        kind: INCOMPATIBLE_ZONE_CP_AND_KUMA_DP_VERSIONS,
        payload: {
          kumaDp: version.kumaDp.version,
        },
      })
    }
  }
}

setWarnings()
</script>

<style lang="scss" scoped>
.reason {
  margin-left: var(--spacing-md);
  margin-bottom: var(--spacing-xxs);
  margin-top: var(--spacing-xxs);
}

.config-wrapper {
  padding-right: var(--spacing-md);
  padding-left: var(--spacing-md);
  padding-bottom: var(--spacing-md);
}
</style>
