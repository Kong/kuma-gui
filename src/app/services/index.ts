import { features } from './features'
import { routes } from './routes'
import { sources } from './sources'
import type { ServiceDefinition } from '@/services/utils'
import { token } from '@/services/utils'

type Token = ReturnType<typeof token>

export const services = (app: Record<string, Token>): ServiceDefinition[] => {
  return [
    [token('services.sources'), {
      service: sources,
      arguments: [
        app.api,
      ],
      labels: [
        app.sources,
      ],
    }],
    [token('services.routes'), {
      service: () => {
        return [routes()]
      },
      labels: [
        app.routes,
      ],
    }],
    [token('services.features'), {
      service: features,
      labels: [
        app.features,
      ],
    }],

  ]
}
