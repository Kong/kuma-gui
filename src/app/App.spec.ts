import { describe, expect, test } from '@jest/globals'
import { flushPromises, mount, RouterLinkStub } from '@vue/test-utils'
import { withVersion } from '@/../jest/jest-setup-after-env'

import App from './App.vue'
import { store } from '@/store/store'
import { TOKENS } from '@/components'
import { set } from '@/services'

function renderComponent(status: string) {
  store.state.globalLoading = true
  store.state.config.tagline = import.meta.env.VITE_NAMESPACE
  store.state.config.status = status

  // keeps the github-button as a <github-button> instead of a span in
  // the snapshot so its as close to actual usage as possible
  set(TOKENS.GithubButton, () => ({
    template: '<github-button />',
    compilerOptions: {
      isCustomElement: () => true,
    },
  }))
  return mount(App, {
    global: {
      stubs: {
        // Letâ€™s not unnecessarily render all that chart markup.
        DonutChart: true,
        'router-link': RouterLinkStub,
      },
    },
  })
}
describe('App.vue', () => {
  test('renders main view when successful', async () => {
    withVersion('10.2.0')
    const wrapper = renderComponent('OK')
    store.dispatch('bootstrap')

    expect(wrapper.find('[data-testid="app-progress-bar"]').exists()).toBe(true)
    await flushPromises()
    expect(wrapper.find('[data-testid="app-progress-bar"]').exists()).toBe(false)

    expect(wrapper.html()).toContain('Create a virtual mesh')
  })

  test('fails to renders basic view', async () => {
    withVersion('10.2.0')
    const wrapper = renderComponent('ERROR')
    store.dispatch('bootstrap')

    expect(wrapper.find('[data-testid="app-progress-bar"]').exists()).toBe(true)
    await flushPromises()
    expect(wrapper.find('[data-testid="app-progress-bar"]').exists()).toBe(false)

    expect(wrapper.element).toMatchSnapshot()
  })
})
