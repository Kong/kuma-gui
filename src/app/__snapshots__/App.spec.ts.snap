// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App.vue fails to renders basic view 1`] = `
<div>
  
  <header
    class="app-header"
  >
    <div
      class="horizontal-list"
    >
      <a>
        <img
          alt="Kuma Logo"
          class="logo-image"
        />
      </a>
      <div
        class="upgrade-check-wrapper"
      >
        <div
          class="upgrade-check"
        >
          <!--v-if-->
        </div>
      </div>
    </div>
    <div
      class="horizontal-list"
    >
      <div
        class="app-status app-status--mobile"
      >
        <div
          aria-controls="aaaabbbb-cccc-dddd-eeee-ffffffffffff"
          data-v-0eaad51f=""
          id="aaaabbbb-cccc-dddd-eeee-ffffffffffff"
          role="button"
        >
          
          <button
            class="medium rounded outline k-button"
            data-v-6f773baa=""
            type="button"
          >
            
            <!---->
            
            
             Info 
            
            <!---->
          </button>
          
          <transition-stub
            appear="false"
            css="true"
            data-v-0eaad51f=""
            name="fade"
            persisted="false"
          >
            <div
              class="k-popover"
              data-v-0eaad51f=""
              id="aaaabbbb-cccc-dddd-eeee-ffffffffffff"
              role="region"
              style="width: 280px; max-width: 350px; max-height: auto; display: none;"
            >
              <!---->
              <div
                class="k-popover-content"
                data-v-0eaad51f=""
              >
                
                <p>
                  Kuma 
                  <b>
                    1.7.1
                  </b>
                   on 
                  <b>
                    Universal
                  </b>
                   (Standalone) 
                </p>
                
              </div>
              <!---->
            </div>
          </transition-stub>
        </div>
      </div>
      <p
        class="app-status app-status--desktop"
      >
        Kuma 
        <b>
          1.7.1
        </b>
         on 
        <b>
          Universal
        </b>
         (Standalone) 
      </p>
      <button
        class="notification-icon cursor-pointer"
        type="button"
      >
        <span
          class="kong-icon-notificationBell kong-icon"
          data-v-71cfcfc4=""
        >
          <svg
            fill="var(--yellow-300)"
            height="24"
            role="img"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            
  
            <title>
              Notifications
            </title>
            
  
            <path
              d="M8.5 19.5a3.8 3.8 0 0 0 3.5 3c1.32 0 3-.97 3.5-3h-7ZM20.42 14.8a.4.4 0 0 0-.11-.26l-1.98-1.97a.5.5 0 0 1-.17-.4v-3.1a5.95 5.95 0 0 0-3.08-5.36c-.3-.18-.63-.31-.95-.46a7.8 7.8 0 0 0-3.96 0l-.2.1c-.93.36-1.73.92-2.4 1.69a5.74 5.74 0 0 0-1.43 3.62c-.03 1.17-.01 2.35-.01 3.52a.5.5 0 0 1-.17.4L4 14.53a.48.48 0 0 0-.13.3V18H20.4l.01-.17c.02-1.01.02-2.02.01-3.03Z"
              fill="var(--yellow-300)"
            />
            

          </svg>
          

        </span>
        <span
          class="notification-icon__amount"
          data-testid="notification-amount"
        >
          4
        </span>
      </button>
      <a>
        <span
          class="kong-icon-gearFilled kong-icon"
          data-v-71cfcfc4=""
        >
          <svg
            height="16"
            role="img"
            viewBox="0 0 16 16"
            width="16"
            xmlns="http://www.w3.org/2000/svg"
          >
            
  
            <title>
              Diagnostics
            </title>
            
  
            <path
              d="M8.63962039 0c.2152152 0 .40628463.13771505.47434165.34188612l.63997656 1.91853291c.3708976.11319954.7265407.26136452 1.0630881.4406537l1.8103874-.9038671c.1924944-.09624716.4249801-.05851992.5771602.09366021l.9045599.90455983c.1521801.15218013.1899073.38466585.0936602.57716018l-.904073 1.81000104c.1795192.33691975.3278441.69298066.44112 1.06432821l1.9182725.63912286C15.862285 6.95409498 16 7.14516441 16 7.36037961v1.27924078c0 .2152152-.137715.40628463-.3418861.47434165l-1.918532.63997355c-.1132072.37092351-.2613848.72659041-.4406907 1.06315871l.9039032 1.8103198c.0962471.1924944.0585199.4249801-.0936602.5771602l-.9045599.9045599c-.1521801.1521801-.3846658.1899073-.5771602.0936602l-1.8103171-.9039046c-.3365681.179306-.692235.3274837-1.0631584.4406912l-.63997656 1.9185329C9.04590502 15.862285 8.85483559 16 8.63962039 16H7.36037961c-.2152152 0-.40628463-.137715-.47434165-.3418861L6.2460614 13.739581c-.3708976-.1131996-.72654071-.2613646-1.06308807-.4406537l-1.81038748.9038671c-.19249433.0962471-.42498005.0585199-.57716018-.0936602l-.90455983-.9045599c-.15218013-.1521801-.18990737-.3846658-.09366021-.5771602l.90407299-1.810001c-.17951921-.3369197-.32784407-.6929806-.44112007-1.0643282L.34188612 9.11396204C.13771505 9.04590502 0 8.85483559 0 8.63962039V7.36037961c0-.2152152.13771505-.40628463.34188612-.47434165l1.918532-.63997355c.11320723-.37092352.26138479-.72659044.44069063-1.06315869l-.90390312-1.81031987c-.09624716-.19249433-.05851992-.42498005.09366021-.57716018l.90455983-.90455983c.15218013-.15218013.38466585-.18990737.57716018-.09366021l1.81031714.90390458c.33656815-.17930601.69223497-.32748375 1.06315841-.44069118L6.88603796.34188612C6.95409498.13771505 7.14516441 0 7.36037961 0h1.27924078zM8 6c-1.1045695 0-2 .8954305-2 2s.8954305 2 2 2 2-.8954305 2-2-.8954305-2-2-2z"
              fill="currentColor"
              fill-rule="evenodd"
            />
            

          </svg>
          

        </span>
        <span
          class="kutil-sr-only"
        >
          Diagnostics
        </span>
      </a>
    </div>
  </header>
  <div
    class="app-content-container"
  >
    <div
      class="app-sidebar-wrapper"
    >
      <aside
        class="app-sidebar"
      >
        
        
        <div
          class="nav-item nav-item--has-bottom-offset"
          data-testid="home"
        >
          <a
            class="nav-link nav-link--is-active"
          >
            Home 
            <!--v-if-->
          </a>
        </div>
        
        
        <div
          class="nav-item nav-item--is-category nav-item--is-primary-category"
          data-testid=""
        >
          <div
            class="nav-category"
          >
            Mesh
          </div>
        </div>
        
        
        
        <div
          class="mesh-selector-container"
        >
          <label
            for="mesh-selector"
          >
            <span
              class="kutil-sr-only"
            >
               Filter by mesh: 
            </span>
            <select
              class="mesh-selector"
              id="mesh-selector"
              name="mesh-selector"
            >
              
              <option
                value="default"
              >
                default
              </option>
              <option
                value="a-test"
              >
                a-test
              </option>
              <option
                value="hello-world"
              >
                hello-world
              </option>
              <option
                value="kong-mania-12"
              >
                kong-mania-12
              </option>
              
            </select>
          </label>
        </div>
        
        
        
        <div
          class="nav-item"
          data-testid="mesh-detail-view"
        >
          <a
            class="nav-link"
          >
            Overview 
            <!--v-if-->
          </a>
        </div>
        
        
        <div
          class="nav-item"
          data-testid="service-list-view"
        >
          <a
            class="nav-link"
          >
            Services 
            <span
              class="amount"
            >
              5
            </span>
          </a>
        </div>
        
        
        <div
          class="nav-item"
          data-testid="data-plane-list-view"
        >
          <a
            class="nav-link"
          >
            Data Plane Proxies 
            <span
              class="amount"
            >
              10
            </span>
          </a>
        </div>
        
        
        <div
          class="nav-item nav-item--is-category nav-item--is-secondary-category"
          data-testid=""
        >
          <div
            class="nav-category"
          >
            Policies
          </div>
        </div>
        
        
      </aside>
    </div>
    <main
      class="app-main-content app-main-content--narrow"
    >
      <!--v-if-->
      <div>
        <div
          class="info icon k-alert mb-4"
          data-testid="notification-info"
          data-v-1910ef98=""
          role="alert"
        >
          <!---->
          <!---->
          <div
            class="k-alert-msg-text has-dismiss-icon"
            data-v-1910ef98=""
          >
            <!---->
            <div
              class="k-alert-msg"
              data-v-1910ef98=""
            >
              
              <div
                class="mr-4"
              >
                <span
                  class="mr-2"
                >
                  <strong>
                    Pro tip:
                  </strong>
                   You might want to adjust your mesh configuration 
                </span>
                <button
                  class="medium rounded outline k-button"
                  data-testid="open-modal-button"
                  data-v-6f773baa=""
                  type="button"
                >
                  
                  <!---->
                  
                  
                   Check your mesh! 
                  
                  <!---->
                </button>
              </div>
              
            </div>
            <!---->
          </div>
          <button
            aria-label="Close"
            class="close"
            data-v-1910ef98=""
            type="button"
          >
            <span
              class="info kong-icon-close kong-icon"
              data-v-1910ef98=""
              data-v-71cfcfc4=""
            >
              <svg
                fill="info"
                height="14"
                role="img"
                viewBox="0 0 20 20"
                width="14"
                xmlns="http://www.w3.org/2000/svg"
              >
                
  
                <title>
                  Close
                </title>
                
  
                <path
                  d="M16 4 3 17m13 0L3 4"
                  stroke="info"
                  stroke-linecap="round"
                  stroke-width="2"
                />
                

              </svg>
              

            </span>
          </button>
          <div
            class="info k-alert-action"
            data-v-1910ef98=""
          >
            <!---->
            <!---->
          </div>
        </div>
        <!---->
      </div>
      <!--v-if-->
      <!--v-if-->
      <transition-stub
        appear="false"
        css="true"
        mode="out-in"
        name="fade"
        persisted="false"
      >
        <div
          class="transition-root"
        >
          TestComponent
        </div>
      </transition-stub>
    </main>
  </div>
  
</div>
`;
